internal/k8sinternal/client.go

```diff
@@ -89,288 +104,108 @@ type ClientOptions struct {
-	for _, statefulSet := range statefulSetList.Items {
-		// For some reason the kubernetes SDK doesn't populate the type meta so we populate it manually
-		statefulSet.TypeMeta = k8s.NewStatefulSet().TypeMeta
-		statefulSets = append(statefulSets, statefulSet.DeepCopyObject())
-	}
-
-	return statefulSets
-}
-
-// GetCronJobs gets all CronJob resources from the cluster
-func GetCronJobs(clientset kubernetes.Interface, options ClientOptions) []k8s.Resource {
-	cronJobClient := clientset.BatchV1beta1().CronJobs(options.Namespace)
-	cronJobList, err := cronJobClient.List(context.Background(), k8s.ListOptionsV1{})
-
-	if err != nil {
-		log.Error(err)
-		return nil
-	}
-
-	cronJobs := make([]k8s.Resource, 0, len(cronJobList.Items))
-	for _, cronJob := range cronJobList.Items {
-		// For some reason the kubernetes SDK doesn't populate the type meta so we populate it manually
-		cronJob.TypeMeta = k8s.NewCronJob().TypeMeta
-		cronJobs = append(cronJobs, cronJob.DeepCopyObject())
-	}
-
-	return cronJobs
-}
-
-// GetNetworkPolicies gets all NetworkPolicy resources from the cluster
-func GetNetworkPolicies(clientset kubernetes.Interface, options ClientOptions) []k8s.Resource {
-	netPolClient := clientset.NetworkingV1().NetworkPolicies(options.Namespace)
-	netPolList, err := netPolClient.List(context.Background(), k8s.ListOptionsV1{})
-
-	if err != nil {
-		log.Error(err)
-		return nil
-	}
-
-	netPols := make([]k8s.Resource, 0, len(netPolList.Items))
-	for _, netPol := range netPolList.Items {
-		// For some reason the kubernetes SDK doesn't populate the type meta so we populate it manually
-		netPol.TypeMeta = k8s.NewNetworkPolicy().TypeMeta
-		netPols = append(netPols, netPol.DeepCopyObject())
-	}
-
-	return netPols
-}
-
-// GetServiceAccounts gets all ServiceAccount resources from the cluster
-func GetServiceAccounts(clientset kubernetes.Interface, options ClientOptions) []k8s.Resource {
-	serviceAccountClient := clientset.CoreV1().ServiceAccounts(options.Namespace)
-	serviceAccountList, err := serviceAccountClient.List(context.Background(), k8s.ListOptionsV1{})
+// GetAllResources gets all supported resources from the cluster
+func (kc kubeClient) GetAllResources(options ClientOptions) []k8s.Resource {
+	var resources []k8s.Resource
 
-	if err != nil {
-		log.Error(err)
-		return nil
+	lists, err := kc.ServerPreferredResources()
+	if err == nil {
+		for _, list := range lists {
+			if len(list.APIResources) == 0 {
+				continue
+			}
+			gv, err := schema.ParseGroupVersion(list.GroupVersion)
+			if err != nil {
+				continue
+			}
+			for _, apiresource := range list.APIResources {
+				if len(apiresource.Verbs) == 0 {
+					continue
+				}
+				gvr := schema.GroupVersionResource{Group: gv.Group, Version: gv.Version, Resource: apiresource.Name}
+
+				// Namespace has to be included as a resource to audit if it is specified.
+				if apiresource.Name == "namespaces" && options.Namespace != "" {
```

Comment:
If I'm understanding correctly, we need to treat namespaces differently because it's a `Get` instead of a `List`? We need the namespace resources for the [netpol audit](https://github.com/Shopify/kubeaudit/blob/main/auditors/netpols/netpols.go#L43) so I think we always want to include the namespace resource. If `options.Namespace != ""` then we only want to include the namespace resource where the namespace name is equal to `options.Namespace`, but it looks like `kc.dynamicClient.Resource(gvr).Get(context.Background(), options.Namespace, metav1.GetOptions{})` takes care of that filtering?
```suggestion
				// Namespace has to be included as a resource to audit if it is specified.
				if apiresource.Name == "namespaces" {
```

Reply:
If the `options.Namespace` is not specified we want to retrieve all namespaces that is why we need to use the [`List`](https://github.com/Shopify/kubeaudit/blob/6e25d12b976b3ac45d0d0ce31866b18a243d233b/internal/k8sinternal/client.go#L155) function instead of the [`Get`](https://github.com/Shopify/kubeaudit/blob/6e25d12b976b3ac45d0d0ce31866b18a243d233b/internal/k8sinternal/client.go#L147) one. 

With the `options.Namespace != ""` test we will use the `Get` will return the namespace according to the option otherwise the `List` will be used to return all namespaces. 

If we use only the `List` that will return an empty unstructuredList  if a namespace is specified:
https://github.com/Shopify/kubeaudit/blob/6e25d12b976b3ac45d0d0ce31866b18a243d233b/internal/k8sinternal/client.go#L155

Reply:
Ohh ok that makes sense. Neat, thank you for the clarification!